net/veth.go

```diff
@@ -0,0 +1,198 @@
+		PeerName: peerName,
+	}
+	if err := netlink.LinkAdd(local); err != nil {
+		return nil, fmt.Errorf(`could not create veth pair %s-%s: %s`, local.Name, local.PeerName, err)
+	}
+
+	cleanup := func(format string, a ...interface{}) (*netlink.Veth, error) {
+		netlink.LinkDel(local)
+		return nil, fmt.Errorf(format, a...)
+	}
+
+	switch maybeBridge.(type) {
+	case *netlink.Bridge:
+		if err := netlink.LinkSetMasterByIndex(local, maybeBridge.Attrs().Index); err != nil {
+			return cleanup(`unable to set master of %s: %s`, local.Name, err)
+		}
+	case *netlink.GenericLink:
+		if maybeBridge.Type() != "openvswitch" {
+			return cleanup(`device "%s" is of type "%s"`, bridgeName, maybeBridge.Type())
+		}
+		if err := odp.AddDatapathInterface(bridgeName, local.Name); err != nil {
+			return cleanup(`failed to attach %s to device "%s": %s`, local.Name, bridgeName, err)
+		}
+	case *netlink.Device:
+		// Assume it's our openvswitch device, and the kernel has not been updated to report the kind.
+		if err := odp.AddDatapathInterface(bridgeName, local.Name); err != nil {
+			return cleanup(`failed to attach %s to device "%s": %s`, local.Name, bridgeName, err)
+		}
+	default:
+		return cleanup(`device "%s" is not a bridge`, bridgeName)
+	}
+
+	if init != nil {
+		guest, err := netlink.LinkByName(peerName)
+		if err != nil {
+			return cleanup("unable to find guest veth %s: %s", peerName, err)
+		}
+		if err := init(local, guest); err != nil {
+			return cleanup("initializing veth: %s", err)
+		}
+	}
+
+	if err := netlink.LinkSetUp(local); err != nil {
+		return cleanup("unable to bring veth up: %s", err)
+	}
+
+	return local, nil
+}
+
+func SetupGuest(guest netlink.Link, name string) error {
+	var err error
+	if err = netlink.LinkSetName(guest, name); err != nil {
+		return err
+	}
+	if err = common.ConfigureARPCache(name); err != nil {
+		return err
+	}
+	return nil
+}
+
+func AddAddresses(guest netlink.Link, cidrs []*net.IPNet) (newAddresses []*net.IPNet, err error) {
+	existingAddrs, err := netlink.AddrList(guest, netlink.FAMILY_V4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get IP address for %q: %v", guest.Attrs().Name, err)
+	}
+	for _, ipnet := range cidrs {
+		if contains(existingAddrs, ipnet) {
+			continue
+		}
+		if err := netlink.AddrAdd(guest, &netlink.Addr{IPNet: ipnet}); err != nil {
+			return nil, fmt.Errorf("failed to add IP address to %q: %v", guest.Attrs().Name, err)
+		}
+		newAddresses = append(newAddresses, ipnet)
+	}
+	return newAddresses, nil
+}
+
+func contains(addrs []netlink.Addr, addr *net.IPNet) bool {
+	for _, x := range addrs {
+		if addr.IP.Equal(x.IPNet.IP) {
+			return true
+		}
+	}
+	return false
+}
+
+func AttachContainer(ns netns.NsHandle, id, ifName, bridgeName string, mtu int, withMulticastRoute bool, cidrs []*net.IPNet) error {
+	// Check if interface already exists
+	if err := common.WithNetNS(ns, func() error {
+		_, err := netlink.LinkByName(ifName)
+		return err
+	}); err != nil {
+		if len(id) > 5 {
```

Comment:
The limit should be the maximum netlink device name length _minus_ "vethwepl".  Which is 7.
I think 5 was picked because originally we were using pids and the longest pid is 5 characters, and we want the names to line up.  But that doesn't seem like a good reason.